ssl/statem/statem_lib.c

```diff
@@ -1515,7 +1516,20 @@ static int is_tls13_capable(const SSL *s)
         default:
             break;
         }
-        if (ssl_has_cert(s, i))
+        if (!ssl_has_cert(s, i))
+            continue;
+        if (i != SSL_PKEY_ECC)
+            return 1;
+        /*
+         * Prior to TLSv1.3 sig algs allowed any curve to be used. TLSv1.3 is
+         * more restrictive so check that our sig algs are consistent with this
+         * EC cert. See section 4.2.3 of RFC8446.
+         */
+        eckey = EVP_PKEY_get0_EC_KEY(s->cert->pkeys[SSL_PKEY_ECC].privatekey);
+        if (eckey == NULL)
+            continue;
+        curve = EC_GROUP_get_curve_name(EC_KEY_get0_group(eckey));
+        if (tls_check_sigalg_curve(s, curve))
```

Comment:
I wonder why this check is ECC specific. The requirements seem to be valid for any certificate type.

Reply:
I guess what I'm proposing is that you make tls_check_sigalg_curve more general.

Reply:
I'm not sure I understand. What would a more general version do? The problem we are trying to solve here is that there are 3 sigalgs defined for TLSv1.2 and below that are also used in TLSv1.3 but have a slightly different meaning. These are:

````
          /* ECDSA algorithms */
          ecdsa_secp256r1_sha256(0x0403),
          ecdsa_secp384r1_sha384(0x0503),
          ecdsa_secp521r1_sha512(0x0603),
````
In TLSv1.2 and below this was defined differently based on the combination of the signature algorithm (0x03 for ecdsa) and the hash (0x04 for sha256, 0x05 for sha384 and 0x06 for sha521). So this gives us:

````
ECDSA+SHA256(0x0403)
ECDSA+SHA384(0x0503)
ECDSA+SHA512(0x0603)
````
So the same numbers have different interpretations in the two protocol versions. Personally I think this is probably a mistake in the TLSv1.3 spec. The difference between the two versions is an additional restriction that TLSv1.3 places on these values to include a specific EC curve. Therefore ECDSA certificates that used to work just fine in TLSv1.2 might suddenly stop working in TLSv1.3 if the client doesn't send a sig alg which is consistent with the curve in the certificate. In the specific case of brainpool certs there are no sigalgs defined for those curves at all - so all servers using brainpool curves in their certs will fail to work in TLSv1.3

This problem is inherently EC specific as it only applies to those three sigalgs where the definition changed.

Reply:
Don't you have the same problem with RSASSA-PSS?

Reply:
But I guess that's not the same kind of problem, and not something you want to check in this function.

Reply:
Exactly.

Reply:
while RSASSA-PSS is different, isn't DSA in the same boat as brainpool? what's its behaviour with TLS 1.3?

Reply:
DSA is already covered by this function (see the beginning of the for loop). If a server is only configured with a DSA cert it will never negotiate TLSv1.3